sympy/physics/quantum/qapply.py

```diff
@@ -1,206 +1,757 @@
+       contradicting results of Mul and qapply.
+       Of course passing factors to Mul makes only sense if there are more than one
+       to pass. So commutative factors are best. So it fits just nicely that in the
+       quantum context the only types Mul knows how to multiply are the scalars, i.e.
+       the complex numbers that may be pulled out.
+
+       So we would best use 'is_scalar' as predicate to send factors to Mul. Unfor-
+       tunately that doesn't exist, and the closest approximation 'is_complex' can
+       not be used: on most symbols, is_complex is undefined. The only predicate that
+       SymPy guarantees is 'is_commutative'!
+
+       So we use 'is_commutative', and we can even profit from args_cnc(). Currently
+       there is no need to filter out commutative multiples of the identity operators,
+       as Identity.is_commutative==False and IdentityOperator.is_commutative==False.
+       But if this quirks should vanish, a better filter is required.
+
+    2. Is the factor a compound type (e.g. OuterProduct, Commutator) that is actually
+       an expression that should be evaluated and broken up into its components to be
+       multiplied one by one, or is it 'elementary'?
+       So qapply implements rules which compound types to break up ((Anti-)Commutator,
+       Pow, Mul, Add) and which to break as an option (OuterProduct). Those that will
+       not be broken up are called "atomic" (note that this is similar in meaning but
+       not identical to the sympy definition of atom/atomic).
+
+       Note that "atomic" refers to the general algorithm (qapply_Mul2) that handles
+       walking the expression tree and the association of factors. qapply still may in-
+       spect the inner contents of factors to do the multiplication (in qapply_Mul2Atoms).
+    """
+
+    def no_interaction(lhs:Expr, rhs:Expr, res:Expr) -> bool:
+        """Returns True if lhs, rhs didn't interact so lhs*rhs just gave res=Mul(lhs,rhs)
+           or Pow(lhs, 2) in case lhs == rhs""" # assumes lhs, rhs non-commutative!
+        return ((res.is_Mul and res.args == (lhs, rhs)) or
+                (lhs == rhs and res.is_Pow and cast(Pow, res).base == lhs))
+
+    # -----------
+    # qapply_Mul2
+    # -----------
+    # Multiply all factors from fL to fR from the left
+    # - fL is typically derived by list(Mul.args), but may contain any expression incl. Mul, Add and c arguments
+    # - fR may not contain Add, Mul or factors Mul should multiply, may only contain "atomic" factors (see above)
+    # - Important: fL, fR are modified during execution!
+
+    def qapply_Mul2(fL:List, fR:List) -> Expr:
+        """Multiplies the left hand list of factors into the right hand list of factors"""
+
+        # define shorthand to distribute eadd: cl*fL*(eadd.args[0]+eadd.args[1]+..)*fR
+        def qapply_DoAddL(eadd:Add): # Important: arguments to qapply_Mul2 must be fresh copies!

```

Comment:
I think it's generally better not to have so many functions and classes defined inside of functions. Here I have to scroll through a lot of stuff to find the actual body of `qapply`.

It's more readable if the functions are short and not nested so I would define these functions outside of `qapply` and just give the name a leading underscore like `_qapply_Mul2`.

Reply:
I believe that is personal taste whether the main function is at the top or at the bottom - I have seen both styles very frequently.
Actually there is only one nesting level, all functions are at same level within qapply.
I also didn't want to use module-wide variables which would be required if I put all functions at module level. So everything is contained within the qapply namespace.

Reply:
This isn't really about personal taste: I found this code hard work to read. There really shouldn't be 700 lines of code between the function signature and the actual body of the function. I had to scan all of those 700 lines to verify that none of them was doing anything more than defining a function or a class before I could confirm that the code at the bottom of the function is the only code that executes here. The fact that most of the lines wrap over in a standard code view (because there are too many characters per line) does not help with this. Also using multiline strings instead of comments does not help because syntax highlighters will colour them differently from comments.

Reply:
Added clear comments where to find start and end of qapply function body and that no function body code is interspersed somewhere. 
So structure and readability are now identical to a module with main function at the bottom.
I analyzed to use the module namespace instead, but would loose important features that come free with using the qapply() namespace as the library wrapper. So I stick to that.